bellman/src/groth16/generator/assembly/eval.rs

```diff
@@ -0,0 +1,205 @@
+                    let mut at = eval_at_tau(powers_of_tau, at);
+                    let mut bt = eval_at_tau(powers_of_tau, bt);
+                    let ct = eval_at_tau(powers_of_tau, ct);
+
+                    // Compute A query (in G1)
+                    if !at.is_zero() {
+                        *a = g1_wnaf.scalar(at.into_repr());
+                    }
+
+                    // Compute B query (in G1/G2)
+                    if !bt.is_zero() {
+                        let bt_repr = bt.into_repr();
+                        *b_g1 = g1_wnaf.scalar(bt_repr);
+                        *b_g2 = g2_wnaf.scalar(bt_repr);
+                    }
+
+                    at.mul_assign(&beta);
+                    bt.mul_assign(&alpha);
+
+                    let mut e = at;
+                    e.add_assign(&bt);
+                    e.add_assign(&ct);
+                    e.mul_assign(inv);
+
+                    *ext = g1_wnaf.scalar(e.into_repr());
+                }
+
+                // Batch normalize
+                E::G1::batch_normalization(a);
+                E::G1::batch_normalization(b_g1);
+                E::G2::batch_normalization(b_g2);
+                E::G1::batch_normalization(ext);
+            });
+        }
+    });
+}
+
+pub struct EvaluationWriter<E>
+where
+    E: Engine
+{
+    pub a: Vec<E::G1>,
+    pub b_g1: Vec<E::G1>,
+    pub b_g2: Vec<E::G2>,
+    pub ic: Option<Vec<E::G1>>,
+    pub l: Vec<E::G1>
+}
+
+impl<E> EvaluationWriter<E> 
+where
+    E: Engine
+{
+    pub fn new(key_pair: &KeyPairAssembly<E>) -> Self {
+        let a = vec![E::G1::zero(); key_pair.num_inputs + key_pair.num_aux];
+        let b_g1 = vec![E::G1::zero(); key_pair.num_inputs + key_pair.num_aux];
+        let b_g2 = vec![E::G2::zero(); key_pair.num_inputs + key_pair.num_aux];
+        let ic = vec![E::G1::zero(); key_pair.num_inputs];
+        let l = vec![E::G1::zero(); key_pair.num_aux];
+        
+        Self { 
+            a, 
+            b_g1, 
+            b_g2, 
+            ic: Some(ic), 
+            l 
+        }
+    }
+
+    pub fn is_unconstrained(&self) -> bool {
+        for e in self.l.iter() {
+            if e.is_zero() {
+                return true
+            }
+        }
+        false
+    }
+    
+    pub fn filter_non_zero_and_map_to_affine(self) -> (Vec<E::G1Affine>, Vec<E::G1Affine>, Vec<E::G1Affine>, Vec<E::G2Affine>) {
+        let l: _ = map_g1_to_affine::<_,E>(self.l);
+        let a: _ = filter_and_map_g1_to_affine::<_,E>(self.a);
+        let b_g1: _ = filter_and_map_g1_to_affine::<_,E>(self.b_g1);
+        let b_g2: _ = filter_and_map_g2_to_affine::<_,E>(self.b_g2);
+            
+        (l, a, b_g1, b_g2)
+    }
+}
+
+fn map_g1_to_affine<I,E>(iter: I) -> Vec<E::G1Affine>
+where
+    I: IntoIterator<Item = E::G1>,
+    E: Engine
+{
+    iter.into_iter()
+        .map(|e| e.into_affine())
+        .collect()
+}
+
+fn filter_and_map_g1_to_affine<I,E>(iter: I) -> Vec<E::G1Affine>
+where
```

Comment:
I'm assuming this is their code, but surely this could be extracted to a single function

Reply:
This is actually our code but I’ve maintained their design pattern, which I’m not overly impressed by. 

I initially tried to run this all as one function, which would have worked if the associated types G1 and G2 both implemented a trait defining the .into_affine() call. I will check again, but they seem to separately require a .into_affine() method. As a consequence, I cannot write a generic function. 

We have two options moving forward:

Option A: I can define a macro that will expand into the functions on compile. 
Option B: I can redefine the interface in the pairing crate so that I can write a generic function. 

I will, if it is a problem, define the macro as a bridge solution until I work on the pairing crate. Otherwise, I will leave this unresolved until a later point.